lgc/patch/PatchReadFirstLane.cpp

```diff
@@ -0,0 +1,245 @@
+    // Collect the instructions that can be marked as uniform
+    DenseSet<Instruction *> markedUniforms(initialReadFirstLanes);
+    // Collect the instructions where the readfirstlanes can be lifted to
+    DenseSet<Instruction *> liftInsts;
+
+    // Reversely iterate over the block to find qualified instructions to lift readfirstlanes
+    for (BasicBlock::reverse_iterator instIt = curBb->rbegin(), endIt = curBb->rend(); instIt != endIt; ++instIt) {
+      Instruction *candidateInst = &*instIt;
+      // The candidate instrction for a lift should be marked as uniform
+      if (markedUniforms.find(candidateInst) != markedUniforms.end()) {
+        unsigned numOps = candidateInst->getNumOperands();
+        if (dyn_cast<CallInst>(candidateInst))
+          --numOps;
+
+        // Stop propagation when it is not worth lifting a readfirstlane
+        if (numOps > 2 || (numOps == 2 && ((m_divergenceAnalysis->isDivergent(candidateInst->getOperand(0)) &&
+                                            m_divergenceAnalysis->isDivergent(candidateInst->getOperand(1)))))) {
+          liftInsts.insert(candidateInst);
+          continue;
+        }
+
+        // Check each operand if it can be marked as uniform or can be qualified as a lift instruction
+        for (unsigned idx = 0; idx < numOps; ++idx) {
+          Value *operand = candidateInst->getOperand(idx);
+          if (auto operandInst = dyn_cast<Instruction>(operand)) {
+            // A readfirstlane lift should occur in one block
+            // Stop propagtion if one of the operands isn't in the same block
+            if (operandInst->getParent() != curBb) {
+              liftInsts.insert(candidateInst);
+              break;
+            }
+
+            // Skip the marked instuction
+            if (markedUniforms.find(operandInst) != markedUniforms.end())
+              continue;
+
+            // Check the operand that cannot be deduced as uniform by divergence analysis
+            if (!m_divergenceAnalysis->isUniform(operandInst)) {
+              // The qualified intruction must has all users marked as uniform
+              bool allUsersUniform = true;
+              for (auto user : operandInst->users()) {
+                auto userInst = dyn_cast<Instruction>(user);
+                if (markedUniforms.find(userInst) == markedUniforms.end()) {
+                  allUsersUniform = false;
+                  break;
+                }
+              }
+              if (allUsersUniform) {
+                auto loadInst = dyn_cast<LoadInst>(operandInst);
+                if (dyn_cast<CallInst>(operandInst) || (loadInst && loadInst->getType()->isIntegerTy(32))) {
+                  // Stop propagation
+                  liftInsts.insert(operandInst);
+                } else {
+                  // Mark the operand as uniform
+                  markedUniforms.insert(operandInst);
+                }
+              }
+            }
+          }
+        }
+      }
+    }
```

Comment:
I have made some modification to this for-loop. the main change is query allUsersMarkedUniform on the def of the current instruction instead of query the operand. because query on the operand may get in-complete result. @xuechen417 @nhaehnle could you help take a look whether this basically looks good? which way do you prefer?
please note that I didn't compile the code, so there may be some build error.
```
/// check whether all operands of \p inst defined in \p block
static bool isAllOperandsDefinedInBlock(Instruction *inst, BasicBlock *block) {
  unsigned numOps = inst->getNumOperands();
  for (unsigned idx = 0; idx < numOps; ++idx) {
    Value *operand = inst->getOperand(idx);
    if (auto operandInst = dyn_cast<Instruction>(operand)) {
      if (operandInst->getParent() != block)
        return false;
    }
  }
  return true;
}
/// Check whether all users of \p inst has been marked as uniform (i.e. been pushed in \p markedUniforms)
static bool isAllUsersMarkedUniform(Instruction *inst, DenseSet<Instruction *> &markedUniforms) {
  for (auto user : inst->users()) {
    auto userInst = dyn_cast<Instruction>(user);
    if (markedUniforms.find(userInst) == markedUniforms.end())
      return false;
  }
  return true;
}
/// a simple heuristic for whether it is beneficial to stop propagate uniform property
bool PatchReadFirstLane::isBeneficialToStop(Instruction *inst) {
  unsigned numOps = inst->getNumOperands();
  if (numOps > 2)
    return true;
  if (numOps == 2 && m_divergenceAnalysis->isDivergent(inst->getOperand(0)) &&
      m_divergenceAnalysis->isDivergent(inst->getOperand(1)))
    return true;
  return false;
}


   BasicBlock::reverse_iterator instIt = curBb->rbegin(), endIt = curBb->rend();
   // skip the instructions we don't care
   for (; instIt != endIt; ++instIt)
      if (MarkedUniforms.find(&*instIt) != MarkedUniforms.end())
        break;

   for (; instIt != endIt; ++instIt) {
      Instruction *candidate = &*instIt;
      if (!m_divergenceAnalysis->isUniform(candidate)) {
        // The candidate instruction is non-uniform, check whether all users of it
        // want to be promoted to uniform
        if (isAllUsersMarkedUniform(candidate, markedUniforms)) {
          auto loadInst = dyn_cast<LoadInst>(candidate);
          // why do we need to stop at i32 type???
          if (isa<CallInst>(candidate) || (loadInst && loadInst->getType()->isIntegerTy(32))) {
            // Stop propagation when the instruction has unknown side-effect
            liftInsts.insert(candidate);
          } else if (isBeneficialToStop(candidate)) {
            // Stop propagation when it is not worth lifting a readfirstlane
            liftInsts.insert(candidate);
          } else if (!isAllOperandsDefinedInBlock(candidate, curBb))
            // stop propagation if some operand is defined in other blocks
            liftInsts.insert(candidate);
          else
            markedUniforms.insert(candidate);
        } else {
          // TODO: check if we can/need insert readfirstlane before the user later.
          for (auto user : candidate->users()) {
            auto userInst = dyn_cast<Instruction>(user);
            if (userInst->getParent() != curBb)
              continue;
            if (markedUniforms.find(userInst) != markedUniforms.end()) {
              liftInsts.insert(userInst);
            }
          }
        }
      }
    }
```

Reply:
The helper functions mostly make sense to me, but I'd tweak the isBeneficialToStop heuristic slightly. I would say: call it "beneficial to stop" if there is more than one non-uniform operand. The reason for this is that you can have a `getelementptr` with many operands. As long as only one of them is divergent, it makes sense to lift the readfirstlane.

You need to stop propagation when you encounter an instruction for which `isSourceOfDivergence` is true: lifting a readfirstlane above such an instruction will give an incorrect result.

I'm still trying to wrap my head around the overall logic of the loop though.

What benefit is there from iterating separately over the basic blocks? Iterating over the instructions of a basic block would make sense if that is your approach for finding `readfirstlane` instructions. However, since you find those instructions via the module-based scan, we could then just rely on a "work queue" approach to look at only the instructions that the algorithm really needs to inspect.

That is, when the algorithm finds a new instruction which is a potential candidate for a readfirstlane being lifted over it, add that instruction to a FIFO. Initialize the FIFO with the initial readfirstlanes. Using a FIFO should ensure that you only need to look at each relevant instruction once. 

Also, I believe the inner loop should be structured as a loop over operands:

```
  Use *divergentUse = nullptr;
  for (Use *use : candidateInst->uses()) {
    if (!m_divergenceAnalysis->isDivergentUse(use))
      continue;
    if (divergentUse) {
      have two divergent uses here: do not lift readfirstlane over candidateInst
    }
    divergentUse = use;
  }
  if (Instruction *divergentDef = dyn_cast<Instruction*>(divergentUse->getValue())) {
    if (divergentDef->getParent() == candidateInst->getParent()) {
      add divergentDef as a new candidate instruction
      continue; // go to the next candidate instruction
    }
  }
  we can still lift the readfirstlane above candidateInst, but we have to make it
  a readfirstlane in the same basic block, and we don't add divergentDef to be
  a new candidate instruction 
  a special case if when divergentUse->getValue() is a function argument; in that case,
  we can put the readfirstlane into the function entry block
```
Does that help? The details will need a bit more thought.

Reply:
Yes, it helps a lot. I'm re-working on this part code as your suggestion.
I have one question that image resource type workaround for gfx10 gives a special case:
```
%.i3 = extractelement <8 x i32> %33, i32 3
...
%39 = icmp sgt i32 %.i3, -1
%40 = and i32 %.i3, 268435455
%41 = select i1 %39, i32 %40, i32 %.i3
%42 = call i32 @llvm.amdgcn.readfirstlane(i32 %.i0)
%43 = call i32 @llvm.amdgcn.readfirstlane(i32 %.i11)
%44 = call i32 @llvm.amdgcn.readfirstlane(i32 %.i2)
%45 = call i32 @llvm.amdgcn.readfirstlane(i32 %41)
```
If we use the stop condition "there is more than one non-uniform operand". We cannot do a lifting.
I discussed with @ruiling and we think this case should be handled. For more than one non-uniform operand, if they can be deduced to one non-uniform, we can lift the readfirstlane over it. What do you think?

Reply:
> The helper functions mostly make sense to me, but I'd tweak the isBeneficialToStop heuristic slightly. I would say: call it "beneficial to stop" if there is more than one non-uniform operand. The reason for this is that you can have a `getelementptr` with many operands. As long as only one of them is divergent, it makes sense to lift the readfirstlane.
> 
The beneficial() function may need further improved, if we prove that we are not inserting more readfirstlane, we are ok to go lift the readfirstlane.

> You need to stop propagation when you encounter an instruction for which `isSourceOfDivergence` is true: lifting a readfirstlane above such an instruction will give an incorrect result.
Yes, that's true.

> 
> I'm still trying to wrap my head around the overall logic of the loop though.
> 
> What benefit is there from iterating separately over the basic blocks? Iterating over the instructions of a basic block would make sense if that is your approach for finding `readfirstlane` instructions. 
You can see my previous explaination under this PR, The reason is try to make sure we have already make sure the users are already visited so we can get correct query result whether all users want to be lifted. 
```
Ins1: %1 = ...
Ins2: %2 = ... %1
Ins3: %3 = ... %1
Ins4: %4 = ... %2
Ins5: %5 = ... %2
Ins6: readfirstlane %3
Ins7: readfirstlane %4
Ins8: readfirstlane %5
```
In the above example, if we want to know whether all users of `%2` want to be optimized as uniform, we need to make sure we have already visited all users of `%2`. So a backward traversal seems a good fit for this, although I admit this would introduce visiting unnecessary instructions :-( Later I realized the current implementation can still get correct query result at the most-closest user because other users already visited at this point.

> However, since you find those instructions via the module-based scan, we could then just rely on a "work queue" approach to look at only the instructions that the algorithm really needs to inspect.

> 
> That is, when the algorithm finds a new instruction which is a potential candidate for a readfirstlane being lifted over it, add that instruction to a FIFO. Initialize the FIFO with the initial readfirstlanes. Using a FIFO should ensure that you only need to look at each relevant instruction once.

> 
> Also, I believe the inner loop should be structured as a loop over operands:
So below part of code are all contained within the loop over operands?
```
for (unsigned idx = 0; idx < numOps; ++idx) {
```
If this is true, may be `candidateInst` means `operandInst`?
> 
> ```
>   Use *divergentUse = nullptr;
>   for (Use *use : candidateInst->uses()) {
>     if (!m_divergenceAnalysis->isDivergentUse(use))
>       continue;
Why using `isDivergentUse()` instead of find in `markedUniform` here? divergentUse means that either the value (candidateInst) is divergent, or it is temporary divergent at the use place. I am not sure what do you want to achieve here?

>     if (divergentUse) {
>       have two divergent uses here: do not lift readfirstlane over candidateInst
>     }
>     divergentUse = use;
>   }
>   if (Instruction *divergentDef = dyn_cast<Instruction*>(divergentUse->getValue())) {
>     if (divergentDef->getParent() == candidateInst->getParent()) {
>       add divergentDef as a new candidate instruction
>       continue; // go to the next candidate instruction
>     }
>   }
>   we can still lift the readfirstlane above candidateInst, but we have to make it
>   a readfirstlane in the same basic block, and we don't add divergentDef to be
>   a new candidate instruction 
>   note that this also covers the special case where divergentUse->getValue()
>   is a function argument
> ```
> 
> Does that help? The details will need a bit more thought.
> 
> Edit: I initially wrote that you can put readfirstlanes into the function entry block if the divergent use value is a function argument. This was incorrect, the readfirstlane must always be in the same basic block as the user.

Reply:
> > What benefit is there from iterating separately over the basic blocks? Iterating over the instructions of a basic block would make sense if that is your approach for finding `readfirstlane` instructions.
> 
> You can see my previous explaination under this PR, The reason is try to make sure we have already make sure the users are already visited so we can get correct query result whether all users want to be lifted.
> 
> ```
> Ins1: %1 = ...
> Ins2: %2 = ... %1
> Ins3: %3 = ... %1
> Ins4: %4 = ... %2
> Ins5: %5 = ... %2
> Ins6: readfirstlane %3
> Ins7: readfirstlane %4
> Ins8: readfirstlane %5
> ```
> 
> In the above example, if we want to know whether all users of `%2` want to be optimized as uniform, we need to make sure we have already visited all users of `%2`. So a backward traversal seems a good fit for this, although I admit this would introduce visiting unnecessary instructions :-( Later I realized the current implementation can still get correct query result at the most-closest user because other users already visited at this point.

I think you make a good point here, actually. I wanted to avoid visiting unnecessary instructions, and I was thinking that by putting instructions into a FIFO in the appropriate way we would be visiting them in the correct order, but I was wrong about that.

Walking the basic block backwards is a simple approach to guarantee correct visit order.

An alternative would be to use an `std::priority_queue` that uses `!Instruction::comesBefore` as its comparison operator. You could add candidates into the priority queue as they become available, and the queue would be able to always give you the last available candidate. The advantage is that it skips over irrelevant instructions. The disadvantage is the increased complexity.

I'd be fine with either solution.


> If this is true, may be `candidateInst` means `operandInst`?

Sorry, in the loop below I should have used `candidateInst->operands()`:

> > ```
> >   Use *divergentUse = nullptr;
> >   for (Use *use : candidateInst->operands()) {
> >     if (!m_divergenceAnalysis->isDivergentUse(use))
> >       continue;
> > ```
> 
> Why using `isDivergentUse()` instead of find in `markedUniform` here? divergentUse means that either the value (candidateInst) is divergent, or it is temporary divergent at the use place. I am not sure what do you want to achieve here?

This is meant to be part of the heuristic for deciding whether to lift readfirstlane above `candidateInst`. Example:
```
  %x = add i32 %a, %b    <---- candidateInst
  %y = call i32 @llvm.amdgcn.readfirstlane(i32 %x)
```
The loop iterates over the uses of `candidateInst`, which are uses of `%a` and of `%b`. If one of them is already known to be uniform, then we don't need to apply readfirstlane to it.

Though there's the problem of the descriptor with fixups. There are two possible ways I can see of addressing this:

1. Ask ourselves why the code comes out this way in the first place. Maybe earlier, when the readfirstlane is first inserted, we can already insert it before the fixup code instead of after it? That would allow you to avoid this case in PatchReadFirstLane.

2. Solve it directly. Do you have a proposal for how that could be done algorithmically?

The most general form of the solution is I believe to first propagate always, and then to figure out all the nodes in the data dependency DAG to which readfirstlane should be applied to minimize the number of readfirstlanes. This problem in turn is a type of minimum cut problem: the nodes to which readfirstlane can be applied form a DAG, and a minimum (node) cut gives you a minimum set of nodes to which to apply readfirstlane. Appropriate weights still need to be added. Min-Cut is a problem that can be solved reasonably fast, but good solvers are non-trivial pieces of code.

With that background, perhaps you're thinking of a convenient heuristic?

Reply:
Yes, I tried a heuristic. It seems work.  To describe clearly, I call a divergent use of `candidateInst` as resultInst.
The available lift is each divergent use of a resultInst is either from resultInst array or from a fixed divergent.
The fixed divergent is from the first resultInst which should have only one divergent use after sorting by comeBefore.
I'm testing the new implementation and will update the PR soon.